# This is the code for the Phase 1 of the Bioinformatics project
# at Sharif University of Technology in Fall Semester of 2022
# By Ali Nazari, Parham Bateni, Masih Najafi
setRepositories() # include number 2 repositories
install.packages(c("BiocManager","GEOquery", "limma", "pheatmap", "ggplot2", "gplots", "reshape2", "plyr", "umap"))
BiocManager::install('GEOquery')
library(GEOquery)

#First we get the gene expression data by finding the correct 
# Series ID in http://www.ncbi.nlm.nih.gov/geo/browse/
# Our desired data is located on https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE48558
Sys.setenv(VROOM_CONNECTION_SIZE=500072)
gset<-getGEO(GEO='GSE48558',GSEMatrix = TRUE, AnnotGPL=TRUE)

# Then we take the desired platform
if (length(gset) > 1) idx <- grep("GPL6244", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]

# Next we extract the expression matrix
ex<-exprs(gset)

AML_numbers=c("GSM1180750", "GSM1180751", "GSM1180752", "GSM1180753", "GSM1180754", "GSM1180755", 
              "GSM1180756", "GSM1180757", "GSM1180758", "GSM1180759", "GSM1180760", "GSM1180761", 
              "GSM1180762", "GSM1180894", "GSM1180895", "GSM1180897", "GSM1180898", "GSM1180899")
Normal_numbers=c("GSM1180790", "GSM1180794", "GSM1180795", "GSM1180818", "GSM1180820", "GSM1180824",
                 "GSM1180826", "GSM1180827", "GSM1180828", "GSM1180829", "GSM1180831", "GSM1180834",
                 "GSM1180835", "GSM1180838", "GSM1180841", "GSM1180843", "GSM1180845", "GSM1180847",
                 "GSM1180849", "GSM1180853", "GSM1180857", "GSM1180887", "GSM1180888", "GSM1180889",
                 "GSM1180890", "GSM1180891", "GSM1180892", "GSM1180893", "GSM1180896", "GSM1180900", 
                 "GSM1180901", "GSM1180902", "GSM1180903", "GSM1180904", "GSM1180905", "GSM1180906",
                 "GSM1180907", "GSM1180908", "GSM1180909", "GSM1180910", "GSM1180911", "GSM1180912",
                 "GSM1180913", "GSM1180914", "GSM1180915", "GSM1180916", "GSM1180917", "GSM1180918",
                 "GSM1180919")

AML_Matrix=ex[, AML_numbers]
Normal_Matrix=ex[, Normal_numbers]

# #################################################################################################################
# these are not our code and are generated by the site which introduced in project file
library(umap)
library(limma)
Sys.setenv(VROOM_CONNECTION_SIZE=500072)
gset<-getGEO(GEO='GSE48558',GSEMatrix = TRUE, AnnotGPL=TRUE)
if (length(gset) > 1) idx <- grep("GPL6244", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]
# make proper column names to match toptable 
fvarLabels(gset) <- make.names(fvarLabels(gset))
# group membership for all samples
gsms <- paste0("1111111111111XXXXXXXXXXXXXXXXXXXXXXXXXXX0XXX0XXXXX",
               "XXXXXXXXXXXXXXXXXX0X0XXX0X0000X0XX00XX00X0X0X0X0X0",
               "XXX0XXX0XXXXXXXXXXXXXXXXXXXXXXXXXXXXX0000000110111",
               "00000000000000000000")
sml <- strsplit(gsms, split="")[[1]]
# filter out excluded samples (marked as "X")
sel <- which(sml != "X")
sml <- sml[sel]
gset <- gset[ ,sel]

ex <- exprs(gset)
qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))
LogC <- (qx[5] > 100) ||
  (qx[6]-qx[1] > 50 && qx[2] > 0)
if (LogC) { ex[which(ex <= 0)] <- NaN
# log2 transformation
exprs(gset) <- log2(ex) }

# assign samples to groups and set up design matrix
gs <- factor(sml)
groups <- make.names(c("Healthy","AML"))
levels(gs) <- groups
gset$group <- gs
design <- model.matrix(~group + 0, gset)
colnames(design) <- levels(gs)
fit <- lmFit(gset, design)  # fit linear model
# set up contrasts of interest and recalculate model coefficients
cts <- paste(groups[1], groups[2], sep="-")
cont.matrix <- makeContrasts(contrasts=cts, levels=design)
fit2 <- contrasts.fit(fit, cont.matrix)
# compute statistics and table of top significant genes
fit2 <- eBayes(fit2, 0.01)
tT <- topTable(fit2, adjust="fdr", sort.by="B", number=250)
tT <- subset(tT, select=c("ID","adj.P.Val","P.Value","t","B","logFC","Gene.symbol","Gene.title"))
write.table(tT, file="GSE48558.top.table.tsv", row.names=F, sep="\t")

# Visualize and quality control test results.
# Build histogram of P-values for all genes. Normal test
# assumption is that most genes are not differentially expressed.
tT2 <- topTable(fit2, adjust="fdr", sort.by="B", number=Inf)
hist(tT2$adj.P.Val, col = "grey", border = "white", xlab = "P-adj",
     ylab = "Number of genes", main = "P-adj value distribution")
# summarize test results as "up", "down" or "not expressed"
dT <- decideTests(fit2, adjust.method="fdr", p.value=0.05)
# Venn diagram of results
vennDiagram(dT, circle.col=palette())
# create Q-Q plot for t-statistic
t.good <- which(!is.na(fit2$F)) # filter out bad probes
qqt(fit2$t[t.good], fit2$df.total[t.good], main="Moderated t statistic")
# volcano plot (log P-value vs log fold change)
colnames(fit2) # list contrast names
ct <- 1        # choose contrast of interest
volcanoplot(fit2, coef=ct, main=colnames(fit2)[ct], pch=20,
            highlight=length(which(dT[,ct]!=0)), names=rep('+', nrow(fit2)))
# MD plot (log fold change vs mean log expression)
# highlight statistically significant (p-adj < 0.05) probes
plotMD(fit2, column=ct, status=dT[,ct], legend=F, pch=20, cex=1)
abline(h=0)

# General expression data analysis
ex <- exprs(gset)
# box-and-whisker plot
dev.new(width=3+ncol(gset)/6, height=5)
ord <- order(gs)  # order samples by group
palette(c("#1B9E77", "#7570B3", "#E7298A", "#E6AB02", "#D95F02",
          "#66A61E", "#A6761D", "#B32424", "#B324B3", "#666666"))
par(mar=c(7,4,2,1))
title <- paste ("GSE48558", "/", annotation(gset), sep ="")
boxplot(ex[,ord], boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=gs[ord])
legend("topleft", groups, fill=palette(), bty="n")
dev.off()
# expression value distribution
par(mar=c(4,4,2,1))
title <- paste ("GSE48558", "/", annotation(gset), " value distribution", sep ="")
plotDensities(ex, group=gs, main=title, legend ="topright")
# UMAP plot (dimensionality reduction)
ex <- na.omit(ex) # eliminate rows with NAs
ex <- ex[!duplicated(ex), ]  # remove duplicates
ump <- umap(t(ex), n_neighbors = 15, random_state = 123)
par(mar=c(3,3,2,6), xpd=TRUE)
plot(ump$layout, main="UMAP plot, nbrs=15", xlab="", ylab="", col=gs, pch=20, cex=1.5)
legend("topright", inset=c(-0.15,0), legend=levels(gs), pch=20,
       col=1:nlevels(gs), title="Group", pt.cex=1.5)
library("maptools")  # point labels without overlaps
pointLabel(ump$layout, labels = rownames(ump$layout), method="SANN", cex=0.6)
# mean-variance trend, helps to see if precision weights are needed
plotSA(fit2, main="Mean variance trend, GSE48558")

# #################################################################################################################
# now we have "adj p.value" and "LogFN" and we continue analysing data with this new file
library(pheatmap)
res = read.delim("GSE48558.top.table.tsv")
aml.up = subset(res, logFC > 1 & adj.P.Val < 0.05)
dim(aml.up)
aml.up.gene = unique(aml.up$Gene.symbol)
length(aml.up.gene)
# now we can check these genes pathways and information
gset <- getGEO("GSE48558", GSEMatrix =TRUE, AnnotGPL=TRUE)
gset_copy = gset
if (length(gset) > 1) idx <- grep("GPL6244", attr(gset, "names")) else idx <- 1
gset <- gset[[idx]]
gsms <- paste0("1111111111111XXXXXXXXXXXXXXXXXXXXXXXXXXX0XXX0XXXXX",
               "XXXXXXXXXXXXXXXXXX0X0XXX0X0000X0XX00XX00X0X0X0X0X0",
               "XXX0XXX0XXXXXXXXXXXXXXXXXXXXXXXXXXXXX0000000110111",
               "00000000000000000000")
sml <- strsplit(gsms, split="")[[1]]
sel <- which(sml != "X")
sml <- sml[sel]
gset <- gset[ ,sel]
ex <- exprs(gset)
normal_cols = which(sml == "1")
cancer_cols = which(sml == "0")
main_normal = gset[ ,normal_cols]
main_cancer = gset[ ,cancer_cols]
main_normal = exprs(main_normal)
main_cancer = exprs(main_cancer)
# let's check normalization
max(ex) # it is 13.8
# so it is in log2 format
min(ex) # it is 1.6
boxplot(ex)
# so we don't need normalization
example_normalization = normalizeQuantiles(ex)
boxplot(example_normalization)
# it is really bad, so we don't need normalization and quality of our data is good
pheatmap(cor(ex))

# #################################################################################################################
pc <- prcomp(ex)
plot(pc, main = "PCA Method", xlab="Dimensions")
plot(pc$x[,1:2], main = "Data on PC1 and PC2 axises", xlab = "PC1", ylab = "PC2")
ex.scale <- t(scale(t(ex), scale = F))
pc.scale <- prcomp(ex.scale)
plot(pc.scale, main = "PCA Method", xlab="Dimensions")
plot(pc.scale$x[,1:2], , main = "Data on PC1 and PC2 axises", xlab = "PC1", ylab = "PC2")

ex_temp <- ex[1:500,]
d <- dist(ex_temp) # euclidean distances between the rows
fit <- cmdscale(d, eig = TRUE, k = 2)
x <- fit$points[, 1]
y <- fit$points[, 2]
plot(x, y, pch = 19, xlim = range(x) + c(0, 600), main = "MDS Method", xlab = "Coordinate1", ylab = "Coordinate2")

install.packages("Rtsne")
gsms <- paste0("1111111111111XXXXXXXXXXXXXXXXXXXXXXXXXXX0XXX0XXXXX",
               "XXXXXXXXXXXXXXXXXX0X0XXX0X0000X0XX00XX00X0X0X0X0X0",
               "XXX0XXX0XXXXXXXXXXXXXXXXXXXXXXXXXXXXX0000000110111",
               "00000000000000000000")
sml <- strsplit(gsms, split="")[[1]]
sel <- which(sml != "X")
sml <- sml[sel]
sml <- t(as.matrix(sml))
IR_data <- ex
IR_species <- sml
library(Rtsne)
tsne_results <- Rtsne(IR_data, perplexity=30, check_duplicates = FALSE)
par(mfrow=c(1,2))
plot(tsne_results$Y, col = "blue", pch = 19, cex = 1.5, main = "tSNE Method", xlab = "tSNE1", ylab = "tSNE2")
plot(tsne_results$Y, col = "black", bg= IR_species, pch = 21, cex = 1.5, main = "tSNE Method", xlab = "tSNE1", ylab = "tSNE2")

# #################################################################################################################
all_amls <- main_normal
all_normals <- main_cancer
granulocytes_list <- c("GSM1180790", "GSM1180794", "GSM1180824", "GSM1180826", "GSM1180887", "GSM1180888", "GSM1180889", "GSM1180890",
                       "GSM1180891", "GSM1180892", "GSM1180893", "GSM1180912")
BCells_list <- c("GSM1180818", "GSM1180829", "GSM1180841", "GSM1180845", "GSM1180900", "GSM1180901", "GSM1180902", "GSM1180903", "GSM1180904",
                 "GSM1180905", "GSM1180906")
TCells_list <- c("GSM1180820", "GSM1180831", "GSM1180834", "GSM1180835", "GSM1180838", "GSM1180839", "GSM1180843", "GSM1180847", "GSM1180896",
                 "GSM1180907", "GSM1180913", "GSM1180914", "GSM1180915", "GSM1180916", "GSM1180917", "GSM1180918", "GSM1180919")
Monocytes_list <- c("GSM1180827", "GSM1180828", "GSM1180908", "GSM1180909", "GSM1180910", "GSM1180911")
CD34_list <- c("GSM1180849", "GSM1180853", "GSM1180857")
aml_group <- all_amls
granulocytes_group <- all_normals[, granulocytes_list]
bcells_group <- all_normals[, BCells_list]
tcells_group <- all_normals[, TCells_list]
monocytes_group <- all_normals[, Monocytes_list]
cd34_group <- all_normals[, CD34_list]
pheatmap(cor(ex))
pheatmap(cor(all_amls))
pheatmap(cor(all_normals))

pdf(file = "comparison.pdf", width = 6, height = 6)
pheatmap(cor(aml_group, granulocytes_group), main = "aml VS. granulocytes")
pheatmap(cor(aml_group, bcells_group), main = "aml VS. bcells")
pheatmap(cor(aml_group, tcells_group), main = "aml VS. tcells")
pheatmap(cor(aml_group, monocytes_group), main = "aml VS. monocytes")
pheatmap(cor(aml_group, cd34_group), main = "aml VS. cd34")
dev.off()
#pheatmap(cor(granulocytes_group, bcells_group), main = "granulocytes VS. bcells")
#pheatmap(cor(granulocytes_group, tcells_group), main = "granulocytes VS. tcells")
#pheatmap(cor(granulocytes_group, monocytes_group), main = "granulocytes VS. monocytes")
#pheatmap(cor(granulocytes_group, cd34_group), main = "granulocytes VS. cd34")

#pheatmap(cor(bcells_group, tcells_group), main = "bcells VS. tcells")
#pheatmap(cor(bcells_group, monocytes_group), main = "bcells VS. monocytes")
#pheatmap(cor(bcells_group, cd34_group), main = "bcells VS. cd34")

#pheatmap(cor(tcells_group, monocytes_group), main = "tcells VS. monocytes")
#pheatmap(cor(tcells_group, cd34_group), main = "tcells VS. cd34")

#pheatmap(cor(monocytes_group, cd34_group), main = "monocytes VS. cd34")
























